This is a guided project from Dataquest.io that explores data from an online auction site for cars. The large bulk of the work was to initially clean the data, and manage the data so that outlier values are not skewing the data. 

After that analysis was done on :
* the popularity of the cars that were listed on the website
* how mileage affects pricing of the cars 
* how in aggregate, damage to cars affects average pricing of cars



In [1]:
import pandas as pd
import numpy as np 
autos = pd.read_csv('autos.csv', encoding = "Latin-1")

In [2]:
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offerType    price   abtest vehicleType  yearOfRegistration  \
0      privat   Angebot   $5,000  control         bus                2004   
1      privat   Angebot   $8,500  control   limousine                1997   
2      privat   Angebot   $8,990     test   limousine                2009   
3      privat   Angebot   $4,350  control  kleinwagen                2007   
4      privat   Angebot   $1,350     test       kombi                2003   
...       ...       ...      ...      ...         ...                 ...   
49995  privat   Angebot  $24,900  control   limousine                2011   
49996  privat   Angebot   $1,980  control      cabrio                1996   
49997  privat   Angebot  $13,200     test      cabrio                2014   
49998  privat   Angebot  $22,900  control       kombi                2013   
49999  privat   Angebot   $1,250  control   limousine                1996   

         gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0        manuell      158  andere  150,000km                    3      lpg   
1      automatik      286     7er  150,000km                    6   benzin   
2        manuell      102    golf   70,000km                    7   benzin   
3      automatik       71  fortwo   70,000km                    6   benzin   
4        manuell        0   focus  150,000km                    7   benzin   
...          ...      ...     ...        ...                  ...      ...   
49995  automatik      239      q5  100,000km                    1   diesel   
49996    manuell       75   astra  150,000km                    5   benzin   
49997  automatik       69     500    5,000km                   11   benzin   
49998    manuell      150      a3   40,000km                   11   diesel   
49999    manuell      101  vectra  150,000km                    1   benzin   

            brand notRepairedDamage          dateCreated  nrOfPictures  \
0         peugeot              nein  2016-03-26 00:00:00             0   
1             bmw              nein  2016-04-04 00:00:00             0   
2      volkswagen              nein  2016-03-26 00:00:00             0   
3           smart              nein  2016-03-12 00:00:00             0   
4            ford              nein  2016-04-01 00:00:00             0   
...           ...               ...                  ...           ...   
49995        audi              nein  2016-03-27 00:00:00             0   
49996        opel              nein  2016-03-28 00:00:00             0   
49997        fiat              nein  2016-04-02 00:00:00             0   
49998        audi              nein  2016-03-08 00:00:00             0   
49999        opel              nein  2016-03-13 00:00:00             0   

       postalCode             lastSeen  
0           79588  2016-04-06 06:45:54  
1           71034  2016-04-06 14:45:08  
2           35394  2016-04-06 20:15:37  
3           33729  2016-03-15 03:16:28  
4           39218  2016-04-01 14:38:50  
...           ...                  ...  
4999

In [3]:
autos.info()
autos.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

Looking through the initial info, we find that there are a few columns that have some null values 

* vehicleType - 44905 non-null values 
* gearbox - 47320 non-null values
* model - 47242 non-null values 
* fuelType - 45518 non-null values
* notRepairedDamage - 40171 non-null values 

In [4]:
column_array = autos.columns
column_array
name_map = {
            'dateCrawled' : 'date_crawled',
            'name': 'name',
            'seller': 'seller',
            'offerType': 'offer_type',
            'price': 'price',
            'abtest': 'ab_test',
            'vehicleType': 'vehicle_type',
            'yearOfRegistration': 'registration_year',
            'gearbox': 'gearbox',
            'powerPS': 'power_ps',
            'model': 'model',
            'odometer': 'odometer',
            'monthOfRegistration': 'registration_month',
            'fuelType': 'fuel_type',
            'brand': 'brand',
            'notRepairedDamage': 'unrepaired_damage',
            'dateCreated': 'ad_created',
            'nrOfPictures': 'nr_of_pictures',
            'postalCode': 'postal_code',
            'lastSeen': 'last_seen'
}

column_array = column_array.map(name_map)
autos.columns = column_array
autos.columns

Index(['date_crawled', 'name', 'seller', 'offer_type', 'price', 'ab_test',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen'],
      dtype='object')

The column names were cleaned to change camelcase to snakecase for readability. Also : 
* yearOfRegistration was changed to registration_year
* monthOfRegistration was changed to registration_month
* notRepairedDamage was changed to unrepaired_damange
* dateCreated was changed to ad_created 

This was done to give better information about each column heading and enhance readability.

In [5]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price  ab_test vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

In [6]:
autos.describe(include = 'all')

date_crawled         name  seller offer_type  price ab_test  \
count                 50000        50000   50000      50000  50000   50000   
unique                48213        38754       2          2   2357       2   
top     2016-03-10 15:36:24  Ford_Fiesta  privat    Angebot     $0    test   
freq                      3           78   49999      49999   1421   25756   
mean                    NaN          NaN     NaN        NaN    NaN     NaN   
std                     NaN          NaN     NaN        NaN    NaN     NaN   
min                     NaN          NaN     NaN        NaN    NaN     NaN   
25%                     NaN          NaN     NaN        NaN    NaN     NaN   
50%                     NaN          NaN     NaN        NaN    NaN     NaN   
75%                     NaN          NaN     NaN        NaN    NaN     NaN   
max                     NaN          NaN     NaN        NaN    NaN     NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  nr_of_pictures   postal_code            last_seen  
count                 50000         50000.0  50000.000000                50000  
unique                   76             NaN           NaN                39481  
top     2016-04-03 00:00:00             NaN           NaN  2016-04-07 06:17:27  
freq                   1946             NaN           NaN                    8  
mean                    NaN             0.0  50813.627300                  NaN  
std                     NaN             0.0  25779.747957                  NaN  
min                     NaN             0.0   1067.000000                  NaN  
25%                     NaN             0.0  30451.000000                  NaN  
50%                     NaN             0.0  49577.000000                  NaN  
75%                     NaN             0.0  71540.000000                  NaN  
max                     NaN             0.0  99998.000000                  NaN

From looking at the data, we find that there are some columns that need investigation 

* vehicle_type = only 8 unique values 
* model - only 245 unique values 

In [7]:
autos.loc[:, "vehicle_type"].value_counts()

limousine     12859
kleinwagen    10822
kombi          9127
bus            4093
cabrio         3061
coupe          2537
suv            1986
andere          420
Name: vehicle_type, dtype: int64

Further investigation into "vehicle_type" shows that this is okay, since it's just classifications of the type of vehicle and thus numbers look ot be valid



In [8]:
autos["model"].value_counts()

golf          4024
andere        3528
3er           2761
polo          1757
corsa         1735
              ... 
kalina           2
i3               1
rangerover       1
200              1
b_max            1
Name: model, Length: 245, dtype: int64

For now, this also looks okay. We have to check price and odometer because these should be numeric values, but they are currently stored as strings

In [9]:
autos["price"].value_counts()

$0          1421
$500         781
$1,500       734
$2,500       643
$1,000       639
            ... 
$299,000       1
$4,840         1
$410           1
$10,090        1
$1,698         1
Name: price, Length: 2357, dtype: int64

In [10]:
price_array = []
for price in autos["price"]:
    price = price.strip()
    price = price.replace("$", "")
    price = price.replace(",", "")
    price_array.append(int(price))
autos['price'] = price_array
autos["price"].value_counts()

0        1421
500       781
1500      734
2500      643
1000      639
         ... 
20790       1
8970        1
846         1
2895        1
33980       1
Name: price, Length: 2357, dtype: int64

Above, we've changed all the values for price into integer values

In [11]:
autos['odometer'].unique()

array(['150,000km', '70,000km', '50,000km', '80,000km', '10,000km',
       '30,000km', '125,000km', '90,000km', '20,000km', '60,000km',
       '5,000km', '100,000km', '40,000km'], dtype=object)

In [12]:
distance_array = []
for distance in autos['odometer']:
    distance = distance.strip()
    distance = distance.replace(',', '')
    distance = distance.replace('km', '')
    distance_array.append(int(distance))
    
autos['odometer'] = distance_array
autos.rename({'odometer': 'odometer_km'}, axis = 1, inplace = True)
autos['odometer_km'].unique()


array([150000,  70000,  50000,  80000,  10000,  30000, 125000,  90000,
        20000,  60000,   5000, 100000,  40000])

In [13]:
autos['seller'].unique().shape



(2,)

In [14]:
autos['offer_type'].unique().shape

(2,)

In [15]:
autos["price"].value_counts().sort_index(ascending = False)

99999999       1
27322222       1
12345678       3
11111111       2
10000000       1
            ... 
5              2
3              1
2              3
1            156
0           1421
Name: price, Length: 2357, dtype: int64

In [16]:
autos['price'].describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64

In [17]:
autos['price'].value_counts().sort_index(ascending = False).head(50)

99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
350000      1
345000      1
299000      1
295000      1
265000      1
259000      1
250000      1
220000      1
198000      1
197000      1
194000      1
190000      1
180000      1
175000      1
169999      1
169000      1
163991      1
163500      1
155000      1
151990      1
145000      1
139997      1
137999      1
135000      1
130000      1
129000      1
128000      1
120000      2
119900      1
119500      1
116000      1
115991      1
115000      1
114400      1
109999      1
105000      2
104900      1
99900       2
99000       2
98500       1
Name: price, dtype: int64

We are dealing with a website where people can list the price of their cars, and cars do vary greatly depending on make and model. However, when looking through the price distributions, we find that : 

* The 75th percentile of car prices listed is only at 7200.00
* The mean price is only at 9840 - and the mean is higher than the 75th percentile

This would signify that there are extremely high price listings at the upper range of pricing that would potentially skew the statistics



In [18]:
print(autos['price'].value_counts().head(20))

0       1421
500      781
1500     734
2500     643
1000     639
1200     639
600      531
800      498
3500     498
2000     460
999      434
750      433
900      420
650      419
850      410
700      395
4500     394
300      384
2200     382
950      379
Name: price, dtype: int64


So a few things to be removed : 
* we will remove all the entries where price is 0
* we will look into removing all entries where price is above 350,000, since price seems to move upwards steadily until there is a large bump in price



In [19]:
new_data = autos[autos['price'].between(1, 350001)]

In [20]:
new_data.describe()

price  registration_year      power_ps    odometer_km  \
count   48565.000000       48565.000000  48565.000000   48565.000000   
mean     5888.935591        2004.755421    117.197158  125770.101925   
std      9059.854754          88.643887    200.649618   39788.636804   
min         1.000000        1000.000000      0.000000    5000.000000   
25%      1200.000000        1999.000000     71.000000  125000.000000   
50%      3000.000000        2004.000000    107.000000  150000.000000   
75%      7490.000000        2008.000000    150.000000  150000.000000   
max    350000.000000        9999.000000  17700.000000  150000.000000   

       registration_month  nr_of_pictures   postal_code  
count        48565.000000         48565.0  48565.000000  
mean             5.782251             0.0  50975.745207  
std              3.685595             0.0  25746.968398  
min              0.000000             0.0   1067.000000  
25%              3.000000             0.0  30657.000000  
50%              6.000000             0.0  49716.000000  
75%              9.000000             0.0  71665.000000  
max             12.000000             0.0  99998.000000

Above we used a **Boolean Index** to select the information relevant if we only keep entries where prices are between 1 and 350000. Here we see that we still have kept around 49k entries, but the mean is now lower that the 75th percentile, which would signify the price statistics are much less skewed towards the upside

In [21]:
autos = new_data

In [22]:
autos['odometer_km'].value_counts().head(20)

150000    31414
125000     5057
100000     2115
90000      1734
80000      1415
70000      1217
60000      1155
50000      1012
5000        836
40000       815
30000       780
20000       762
10000       253
Name: odometer_km, dtype: int64

In [23]:
autos['odometer_km'].value_counts().sort_index(ascending = False).head(20)

150000    31414
125000     5057
100000     2115
90000      1734
80000      1415
70000      1217
60000      1155
50000      1012
40000       815
30000       780
20000       762
10000       253
5000        836
Name: odometer_km, dtype: int64

In [24]:
autos['odometer_km'].describe()

count     48565.000000
mean     125770.101925
std       39788.636804
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

We find that the kilometers logged in the odometer are rounded - but this could be due to the fact that it's an online auction website, so it's possible that the information is being rounded due to the way the website forces entry. 

Also, out of the 48565 entries remaining after adjusting for price outliers, 31414 entries have logged over 150000km on the odometer. This translates to **64%** of all cars being at the extreme high range of kilometers logged. It could be possible that the website only allows for a max of 150000, and since it seems like these cars listed are second hand, that a lot of people who have logged a lot of mileage on their cars are selling on this website

In [25]:
autos['odometer_km'].value_counts().sort_index(ascending = True).head(20)


5000        836
10000       253
20000       762
30000       780
40000       815
50000      1012
60000      1155
70000      1217
80000      1415
90000      1734
100000     2115
125000     5057
150000    31414
Name: odometer_km, dtype: int64

Since the minimum distance logged is 5000, we can see that it seems that there are no outliers in the odometer_km column that we should remove and all statistics seem valid.

In [26]:
autos.head(5)

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type  price  ab_test vehicle_type  registration_year  \
0  privat    Angebot   5000  control          bus               2004   
1  privat    Angebot   8500  control    limousine               1997   
2  privat    Angebot   8990     test    limousine               2009   
3  privat    Angebot   4350  control   kleinwagen               2007   
4  privat    Angebot   1350     test        kombi               2003   

     gearbox  power_ps   model  odometer_km  registration_month fuel_type  \
0    manuell       158  andere       150000                   3       lpg   
1  automatik       286     7er       150000                   6    benzin   
2    manuell       102    golf        70000                   7    benzin   
3  automatik        71  fortwo        70000                   6    benzin   
4    manuell         0   focus       150000                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

In [27]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 48565 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   date_crawled        48565 non-null  object
 1   name                48565 non-null  object
 2   seller              48565 non-null  object
 3   offer_type          48565 non-null  object
 4   price               48565 non-null  int64 
 5   ab_test             48565 non-null  object
 6   vehicle_type        43979 non-null  object
 7   registration_year   48565 non-null  int64 
 8   gearbox             46222 non-null  object
 9   power_ps            48565 non-null  int64 
 10  model               46107 non-null  object
 11  odometer_km         48565 non-null  int64 
 12  registration_month  48565 non-null  int64 
 13  fuel_type           44535 non-null  object
 14  brand               48565 non-null  object
 15  unrepaired_damage   39464 non-null  object
 16  ad_created          48

Because **date_crawled** , **ad_created** and **last_seen** are all stored as strings, we aren't able to get proper distribution details for these columns as they are stored in datetime stamps

In [28]:
autos[["date_crawled", 'ad_created', "last_seen"]].head()

date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

As all columns have 0 null values, we can safely extract just the dates from the datetime values, and then reassign them back to their individual columns

In [29]:
(autos['date_crawled'].str[:10].value_counts
 (normalize = True, dropna = False).sort_index())

2016-03-05    0.025327
2016-03-06    0.014043
2016-03-07    0.036014
2016-03-08    0.033296
2016-03-09    0.033090
2016-03-10    0.032184
2016-03-11    0.032575
2016-03-12    0.036920
2016-03-13    0.015670
2016-03-14    0.036549
2016-03-15    0.034284
2016-03-16    0.029610
2016-03-17    0.031628
2016-03-18    0.012911
2016-03-19    0.034778
2016-03-20    0.037887
2016-03-21    0.037373
2016-03-22    0.032987
2016-03-23    0.032225
2016-03-24    0.029342
2016-03-25    0.031607
2016-03-26    0.032204
2016-03-27    0.031092
2016-03-28    0.034860
2016-03-29    0.034099
2016-03-30    0.033687
2016-03-31    0.031834
2016-04-01    0.033687
2016-04-02    0.035478
2016-04-03    0.038608
2016-04-04    0.036487
2016-04-05    0.013096
2016-04-06    0.003171
2016-04-07    0.001400
Name: date_crawled, dtype: float64

Looking at the dates of **date_crawled** , we can see that the distribution of dates crawed across the year is fairly even. 

In [30]:
(autos['last_seen'].str[:10].value_counts(normalize=True, dropna = False)
 .sort_index()
)

2016-03-05    0.001071
2016-03-06    0.004324
2016-03-07    0.005395
2016-03-08    0.007413
2016-03-09    0.009595
2016-03-10    0.010666
2016-03-11    0.012375
2016-03-12    0.023783
2016-03-13    0.008895
2016-03-14    0.012602
2016-03-15    0.015876
2016-03-16    0.016452
2016-03-17    0.028086
2016-03-18    0.007351
2016-03-19    0.015834
2016-03-20    0.020653
2016-03-21    0.020632
2016-03-22    0.021373
2016-03-23    0.018532
2016-03-24    0.019767
2016-03-25    0.019211
2016-03-26    0.016802
2016-03-27    0.015649
2016-03-28    0.020859
2016-03-29    0.022341
2016-03-30    0.024771
2016-03-31    0.023783
2016-04-01    0.022794
2016-04-02    0.024915
2016-04-03    0.025203
2016-04-04    0.024483
2016-04-05    0.124761
2016-04-06    0.221806
2016-04-07    0.131947
Name: last_seen, dtype: float64

The crawler records last_seen based on the last time it had seen the listing for the car, which lets us know on what day a listing was removed. The last three days show a a spime in last seen values, but given that it's unlikely there was a large spike in sales, it's more likely that this is due to the cutoff in crawling dates.

In [31]:
(
autos['ad_created'].str[:10]
    .value_counts(normalize = True, dropna = False)
    .sort_index()
)


2015-06-11    0.000021
2015-08-10    0.000021
2015-09-09    0.000021
2015-11-10    0.000021
2015-12-05    0.000021
                ...   
2016-04-03    0.038855
2016-04-04    0.036858
2016-04-05    0.011819
2016-04-06    0.003253
2016-04-07    0.001256
Name: ad_created, Length: 76, dtype: float64

There is a large variety of ad created dates. Most fall within 1-2 months of the listing date, but a few are quite old, with the oldest at around 9 months.


In [32]:
autos['registration_year'].describe()

count    48565.000000
mean      2004.755421
std         88.643887
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

The registration year is likely tied to the age of the car
Looking through the values, we find that a max of 9999 and a min of 1000 are present, where both values don't make sense and these are likely erroneous values

In [33]:
(~autos["registration_year"].between(1900,2016)).sum() / autos.shape[0]


0.038793369710697

Here we can see that if we remove values from below 1900 and above 2016, we would then only be removing less than 4% of the values in totality so we will move forward to do so.

In [34]:
autos = autos[autos["registration_year"].between(1900, 2016)]

In [35]:
autos['registration_year'].describe()

count    46681.000000
mean      2002.910756
std          7.185103
min       1910.000000
25%       1999.000000
50%       2003.000000
75%       2008.000000
max       2016.000000
Name: registration_year, dtype: float64

Here we can see that instead of **48565** values, we have **46681** values remaining, which shows we've successfully removed all the values that fall outside our criteria for **registration year**

We will explore all brands that take up more than 5% of all listings on the website

In [70]:
brands = autos['brand'].value_counts(normalize = True, dropna = False)
brands

volkswagen        0.211264
bmw               0.110045
opel              0.107581
mercedes_benz     0.096463
audi              0.086566
ford              0.069900
renault           0.047150
peugeot           0.029841
fiat              0.025642
seat              0.018273
skoda             0.016409
nissan            0.015274
mazda             0.015188
smart             0.014160
citroen           0.014010
toyota            0.012703
hyundai           0.010025
sonstige_autos    0.009811
volvo             0.009147
mini              0.008762
mitsubishi        0.008226
honda             0.007840
kia               0.007069
alfa_romeo        0.006641
porsche           0.006127
suzuki            0.005934
chevrolet         0.005698
chrysler          0.003513
dacia             0.002635
daihatsu          0.002506
jeep              0.002271
subaru            0.002142
land_rover        0.002099
saab              0.001649
jaguar            0.001564
daewoo            0.001500
trabant           0.001392
r

We can see that in terms of percentages, the listings are clustered towards a certain number of brands so for this analysis we will be looking at brands that take up more than 4% of all listings on the website

In [75]:
common_brands = brands[brands > 0.04].index
common_brands

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford',
       'renault'],
      dtype='object')

In [76]:
price_index = {}
for brand in common_brands: 
    price_index[brand] = \
    autos.loc[autos["brand"] == brand, "price"].mean()
price_index

{'volkswagen': 5402.410261610221,
 'bmw': 8332.820517811953,
 'opel': 2975.2419354838707,
 'mercedes_benz': 8628.450366422385,
 'audi': 9336.687453600594,
 'ford': 3749.4695065890287,
 'renault': 2474.8646069968195}

In [78]:
mileage_index = {}
for brand in common_brands:
    mileage_index[brand] = \
    autos.loc[autos['brand'] == brand, "odometer_km"].mean()
mileage_index

{'volkswagen': 128707.15879132022,
 'bmw': 132572.51313996495,
 'opel': 129310.0358422939,
 'mercedes_benz': 130788.36331334666,
 'audi': 129157.38678544914,
 'ford': 124266.01287159056,
 'renault': 128071.33121308497}

In [79]:
price_series = pd.Series(price_index)
mileage_series = pd.Series(mileage_index)
dataframe = pd.DataFrame(price_series, columns = ["mean_price"])
dataframe["mean_mileage"] = mileage_series

In [80]:
dataframe

mean_price   mean_mileage
volkswagen     5402.410262  128707.158791
bmw            8332.820518  132572.513140
opel           2975.241935  129310.035842
mercedes_benz  8628.450366  130788.363313
audi           9336.687454  129157.386785
ford           3749.469507  124266.012872
renault        2474.864607  128071.331213

In [81]:
dataframe.sort_values('mean_price', ascending = False)

mean_price   mean_mileage
audi           9336.687454  129157.386785
mercedes_benz  8628.450366  130788.363313
bmw            8332.820518  132572.513140
volkswagen     5402.410262  128707.158791
ford           3749.469507  124266.012872
opel           2975.241935  129310.035842
renault        2474.864607  128071.331213

Above, we've put together a dataframe that shows :
* mean price of all cars listed that make up more than 4% of listings
* mean mileage of all cars listed that make up more than 4% of listings 

In [82]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type  price  ab_test vehicle_type  registration_year  \
0  privat    Angebot   5000  control          bus               2004   
1  privat    Angebot   8500  control    limousine               1997   
2  privat    Angebot   8990     test    limousine               2009   
3  privat    Angebot   4350  control   kleinwagen               2007   
4  privat    Angebot   1350     test        kombi               2003   

     gearbox  power_ps   model  odometer_km  registration_month fuel_type  \
0    manuell       158  andere       150000                   3       lpg   
1  automatik       286     7er       150000                   6    benzin   
2    manuell       102    golf        70000                   7    benzin   
3  automatik        71  fortwo        70000                   6    benzin   
4    manuell         0   focus       150000                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

Now we look to translate categorical words that are in german to english for this data. The categories we should be looking for are : 

* gearbox
* fuel_type
* unrepaired damage

In [83]:
gearbox_series = autos['gearbox']
gearbox_series.unique()

array(['manuell', 'automatik', nan], dtype=object)

In [84]:
gear_map = {
    'manuell': 'manual',
    'automatik': 'automatic'
}
gearbox_series = gearbox_series.map(gear_map)

In [86]:
autos['gearbox'] = gearbox_series
autos['gearbox'].unique()

<ipython-input-86-83ee7044c5a7>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  autos['gearbox'] = gearbox_series


array(['manual', 'automatic', nan], dtype=object)

In [87]:
autos['fuel_type'].unique()

array(['lpg', 'benzin', 'diesel', nan, 'cng', 'hybrid', 'elektro',
       'andere'], dtype=object)

In [89]:
fuel_map = {
    'lpg': 'lpg',
    'benzin': 'petrol',
    'diesel': 'diesel',
    'cng': 'cng',
    'hybrid': 'hybrid',
    'elektro': 'electric',
    'andere': 'other'
}
autos['fuel_type'] = autos['fuel_type'].map(fuel_map)

<ipython-input-89-c2781aa7ef49>:10: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  autos['fuel_type'] = autos['fuel_type'].map(fuel_map)


In [90]:
autos['unrepaired_damage'].unique()

array(['nein', nan, 'ja'], dtype=object)

In [91]:
repair_map = {
    'nein': 'no',
    'ja': 'yes'
}
autos['unrepaired_damage'] = \
autos['unrepaired_damage'].map(repair_map)

<ipython-input-91-aeddccd7a1fd>:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  autos['unrepaired_damage'] = \


In [93]:
autos[['fuel_type', 'unrepaired_damage', 'gearbox']].head()

fuel_type unrepaired_damage    gearbox
0       lpg                no     manual
1    petrol                no  automatic
2    petrol                no     manual
3    petrol                no  automatic
4    petrol                no     manual

We've now sorted through all the german words, and changed them to English equivalents

In [95]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 46681 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   date_crawled        46681 non-null  object
 1   name                46681 non-null  object
 2   seller              46681 non-null  object
 3   offer_type          46681 non-null  object
 4   price               46681 non-null  int64 
 5   ab_test             46681 non-null  object
 6   vehicle_type        43977 non-null  object
 7   registration_year   46681 non-null  int64 
 8   gearbox             44571 non-null  object
 9   power_ps            46681 non-null  int64 
 10  model               44488 non-null  object
 11  odometer_km         46681 non-null  int64 
 12  registration_month  46681 non-null  int64 
 13  fuel_type           43363 non-null  object
 14  brand               46681 non-null  object
 15  unrepaired_damage   38374 non-null  object
 16  ad_created          46

Datetimes would be stored in :
* date_crawled
* last_seen 

In [97]:
autos['date_crawled'].unique()

array(['2016-03-26 17:47:46', '2016-04-04 13:38:56',
       '2016-03-26 18:57:24', ..., '2016-03-28 10:50:25',
       '2016-03-08 19:25:42', '2016-03-14 00:42:12'], dtype=object)

For these dates, we now want to store them as unique integers so that we would be able to sort through them using booleans

In [100]:
date_crawled_list = []
for entry in autos['date_crawled']:
    date = entry[:10]
    date = date.strip().replace('-', '')
    date_crawled_list.append(date)

In [102]:
autos['date_crawled'] = date_crawled_list

<ipython-input-102-c40e2c099d2b>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  autos['date_crawled'] = date_crawled_list


In [104]:
last_seen_list = []
for entry in autos['last_seen']:
    date = entry[:10]
    date = date.strip().replace('-', '')
    last_seen_list.append(date)

autos.loc[:, 'last_seen'] = last_seen_list

/dataquest/system/env/python3/lib/python3.8/site-packages/pandas/core/indexing.py:966: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self.obj[item] = s


In [105]:
autos[["date_crawled", 'last_seen']].head()

date_crawled last_seen
0     20160326  20160406
1     20160404  20160406
2     20160326  20160406
3     20160312  20160315
4     20160401  20160401

In [106]:
autos["name"].unique()

array(['Peugeot_807_160_NAVTECH_ON_BOARD',
       'BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik',
       'Volkswagen_Golf_1.6_United', ...,
       'Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon',
       'Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+Reifen_neu_!!',
       'Fiat_500_C_1.2_Dualogic_Lounge'], dtype=object)

In [109]:
print(autos.columns)

Index(['date_crawled', 'name', 'seller', 'offer_type', 'price', 'ab_test',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer_km', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen'],
      dtype='object')


In [110]:
autos['brand_model'] = autos['brand'] + "|" + autos['model']

<ipython-input-110-3eacee9c7b5a>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  autos['brand_model'] = autos['brand'] + "|" + autos['model']


In [112]:
autos['brand_model'].value_counts().head()

volkswagen|golf      3707
bmw|3er              2615
volkswagen|polo      1609
opel|corsa           1592
volkswagen|passat    1349
Name: brand_model, dtype: int64

|Brand | Model |
|:---: | :---:|
|Volkswagen | Golf|
|BMW | 3er|
|Volkswagen | Polo|
|Opel | Corsa|
|Volkswagen | Passat|


The most common brand/model combinations are shown above 

In [115]:
autos['odometer_km'].describe()

count     46681.000000
mean     125586.855466
std       39852.528628
min        5000.000000
25%      100000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

In [130]:
print("Average price of cars mileage up to 30000 miles")
print(autos.loc[autos['odometer_km'] < 30000, \
                'price'].mean().round(2))
print("Between 30001 and 60000")
print(autos.loc[autos['odometer_km'].between(30001, 60000), \
                'price'].mean().round(2))
print('Between 60001 and 90000')
print(autos.loc[autos['odometer_km'].between(60001, 90000), \
                'price'].mean().round(2))
print('Between 90001 and 120000')
print(autos.loc[autos['odometer_km'].between(90001, 120000), \
                'price'].mean().round(2))
print('Between 120001 and 150000')
print(autos.loc[autos['odometer_km'].between(120001, 150000), \
                'price'].mean().round(2))


Average price of cars mileage up to 30000 miles
14483.73
Between 30001 and 60000
13721.36
Between 60001 and 90000
9563.22
Between 90001 and 120000
8132.7
Between 120001 and 150000
4107.94


We see that for cars up to 60000 km in mileage, the drop in average price isn't very much affected. However, past the 60000km mark, average price drops quite substantially as compared to before, and then there is another large price drop at the 120000km mark 

In [135]:
print('Average price of non-damaged cars')
print(autos.loc[autos['unrepaired_damage'] == 'no', "price"].mean())
print('Average price of damaged cars')
print(autos.loc[autos['unrepaired_damage'] == 'yes', 'price'].mean())

Average price of non-damaged cars
7164.033102796004
Average price of damaged cars
2241.146035242291


In [137]:
(7164 - 2241) / 7164

0.6871859296482412

We see that on average, having unrepaired damage on a car can cause the price to be dropped as low as 68% lower fr